### ISSUE TITLE
move to typed function declarations (dropping support for python2)

### ISSUE THREAD

**Issue by dkg** [SecurityInnovation/PGPy#300]
[python 2 is reaching EOL in a few months](https://github.com/python/devguide/pull/344).

If we drop support for python2 in PGPy, then we can take full advantage of a number of features that are possible only in python3, like [explicit type annotations in function declarations](https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html#functions).  If we add these annotations, then we can also remove the type information from the docstrings, because they'll already be documented.

I'd be willing to try to add these annotations and clean up the docstrings, but i don't want to do it if these annotations need to be in the [much clumsier python2 style](https://mypy.readthedocs.io/en/latest/cheat_sheet.html#functions).

How committed are we to supporting python2 in future versions of PGPy?

**Comment by Commod0re** [comment]
I think we should target dropping support for Py2 soon. I intend to add one
or two more things over the weekend, then publish the next release next
week, and then we should work on dropping support for py2. I've been
waiting years to say that 😆

**Comment by dkg** [comment]
@Commod0re i don't think that next release was ever published.  can you go through with that so that we can start looking at what the next steps might be?
